cpp/CMakeLists.txt

```diff
@@ -49,6 +49,11 @@ option(DISABLE_DEPRECATION_WARNINGS "Disable depreaction warnings " ON)
 option(DISABLE_OPENMP "Disable OpenMP" OFF)
 option(NVTX "Enable nvtx markers" OFF)
 
+option(RAFT_COMPILE_LIBRARIES "Enable building raft shared library instantiations" ON)
+option(RAFT_ENABLE_NN_DEPENDENCIES "Search for raft::nn dependencies like faiss" ${RAFT_COMPILE_LIBRARIES})
+include(CMakeDependentOption)
+cmake_dependent_option(RAFT_USE_FAISS_STATIC "Build and statically link the FAISS library for nearest neighbors search on GPU" ON RAFT_COMPILE_LIBRARIES OFF)
```

Comment:
I opened a [PR](https://github.com/rapidsai/cuml/pull/4476) w/ cuml to integrate the changes in this PR but it looks like it's trying to build FAISS from source every time now, rather than using the shared lib that's already installed w/ conda. 

I also tried removing the explicit call to `get_faiss` and linking  of the `FAISS::FAISS` targets in cuml but it didn't seem to make a difference.

Reply:
Your PR doesn't change the version of RAFT it is using to this PR.

Reply:
Ah, good catch. I've been building locally w/ `CPM_raft_SOURCE`, though. Let me update the PR as well.

Reply:
[Here](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cuml/job/prb/job/cuml-cpu-cuda-build-arm64/CUDA=11.5/518/console) it looks like it's building FAISS from source in CI in cuml.

Reply:
This PR changes the name of the searched package to `faiss` ( lowercase ) to follow modern CMake patterns, while your cuml PR keeps the existing behavior of using `FAISS`.

This for CPM means they are distinctly different projects. When we look at the build logs we see the following:
```
-- CPM: using local package FAISS@1.7.0
-- CPM: raft: adding package faiss@1.7.0 (bde7c0027191f29c9dadafe4f6e68ca0ee31fb30)
```

So we can see that `FAISS` was found locally, and `faiss` is going to be built from source. 
In general we wouldn't expect RAFT search again for `faiss` but given the mismatch on names it does so. 

As for why RAFT is building `faiss` from source, the cuml `get_raft.cmake` project file currently enables `CPM_DOWNLOAD_ALL` which forces raft and all dependencies to be built from source and not search for a local copy.

So what can you do?

1. Update cuml to search for `faiss` instead of `FAISS`
2. Don't have `get_raft` in cuml always trying to build `raft` from source. Since this will also stop the benifits of `raft::nn` and `raft::distance`

Reply:
I've verified (at least locally) that changing the faiss target (and `GLOBAL_TARGETS` for faiss to `faiss::faiss`) works when both with and without `CPM_raft_SOURCE`. 

We're working towards item number 2. conda packaging will make that much easier to implement and maintain.